File: algorithms/Java/linked-lists/circular.java
class CircularLinkedList{
    class Node {
        int value;
        Node next;

        public Node(int value) {
            this.value = value;
        }
    }

    private Node head = null;
    private Node tail = null;

    public void addNode(int value) {
        Node newNode = new Node(value);

        if (head == null) {
            head = newNode;
        } else {
            tail.next = newNode;
        }

        tail = newNode;
        tail.next = head;
    }
    public void printNodes() {  
        Node current = head;  
        if(head == null) {  
            System.out.println("List is empty");
            return;  
        }    
        System.out.println("The circular linked list is: "); 
        System.out.print("[ "); 
        do{   
            System.out.print(current.value + " ");  
            current = current.next;
            //go to next until we get back to the head 
        }while(current != head);
        System.out.println("]"); 
    }
}

public class circular {
    public static void main(String [] args){
        CircularLinkedList cll = new CircularLinkedList();

        cll.addNode(10);
        cll.addNode(12);
        cll.addNode(2);
        cll.addNode(1);
        cll.addNode(8);
        cll.addNode(42);
        cll.addNode(17);

        cll.printNodes();

        //this will print :
        /*
            The circular linked list is: 
            [ 10 12 2 1 8 42 17 ]
        */
    }
}

/*
    to run the file:
    javac circular.java
    java circular
*/

File: algorithms/C++/Linked-Lists/circular.cpp
//Description : Creating Circular linked list

#include <iostream>
using namespace std;

struct Node {
    int data;
    Node * next;
    Node(int x) {
        data = x;
        next = NULL;
    }
};
Node * insertion(int newdata, Node * head) {  //Insertion at the end of linked list
    Node * newnode = new Node(newdata);       //Creating a new node with value newdata
    Node * ptr = head;
    if (head == NULL) {                       //If the list was initially empty then return the newnode as the new head 
        newnode -> next = newnode;
        return newnode;
    }
    while (ptr -> next != head) {            //Traversing till we reach the last node
        ptr = ptr -> next;
    }
    newnode -> next = ptr -> next;          //Insert the new node at the end of the list
    ptr -> next = newnode;
    return head;

}

void display(Node * head) {                //Function to print items in the list
    Node * ptr = head;
    if (head == NULL) {
        cout << "Empty Circular Linked List";
        return;
    }
    do {
        cout << ptr -> data << " ";
        ptr = ptr -> next;
    }
    while (ptr != head);
}

int main() {                            //Main Driver function
    Node * head = NULL;
    head = insertion(3, head);
    head = insertion(4, head);
    head = insertion(5, head);
    head = insertion(6, head);
    head = insertion(1, head);
    head = insertion(7, head);
    display(head);
    return 0;

}
